functionary/functionary/settings/base/auth_.py

```diff
@@ -1,15 +1,61 @@
-SITE_ID = 1
-
 ACCOUNT_ADAPTER = "ui.admin.auth.FunctionaryAccountAdapter"
 ACCOUNT_AUTHENTICATION_METHOD = "username"
 ACCOUNT_EMAIL_VERIFICATION = "none"
 ACCOUNT_LOGIN_ATTEMPTS_TIMEOUT = 900
 ACCOUNT_PRESERVE_USERNAME_CASING = False
 ACCOUNT_SESSION_REMEMBER = True
-
+USER_MODEL_EMAIL_FIELD = None
 AUTHENTICATION_BACKENDS = (
     "django.contrib.auth.backends.ModelBackend",
     "allauth.account.auth_backends.AuthenticationBackend",
 )
 
-SOCIALACCOUNT_PROVIDERS = {}
+CONSTANCE_BACKEND = "constance.backends.database.DatabaseBackend"
+CONSTANCE_DATABASE_PREFIX = "constance:functionary:"
+CONSTANCE_ADDITIONAL_FIELDS = {"config_field": ["django.forms.fields.JSONField", {}]}
+CONSTANCE_CONFIG = {
+    "SOCIALACCOUNT_PROVIDERS": (
+        {
+            "github": {},
+            "gitlab": {},
+            "keycloak": {},
+        },
+        "SocialAccount Providers Configuration",
+        "config_field",
+    ),
+}
+
+
+def constance_settings_proxy(setting_name, default_value):
+    """Custom settings function for django-allauth.
+
+    This function will check Constance for the given setting_name and
+    revert to the django.conf.settings if it's not found. This function
+    allows the SOCIALACCOUNT_PROVIDERS to be configured outside of these
+    config files.
+
+    Args:
+        setting_name: The name of the setting to fetch
+        default_value: The default to use if it's not configured
+
+    Returns:
+        The configured setting from Constance. If it's not found, the setting
+        from django.conf.settings, otherwise the default_value.
+    """
+    import json
+
+    from constance import config
+    from django.conf import settings
+
+    value = getattr(config, setting_name, None)
+    try:
```

Comment:
This was created for the `SOCIALACCOUNT_PROVIDERS` property, but I didn't want to restrict it to just that one property.

It first attempts to get the value from Constance. It will then attempt to parse the value as json. If it's not json, the value is still what was loaded from Constance.

The last `getattr()` is the default django-allauth implementation of the property getter. I changed the if to check for `value is None` instead of `not value` to preserve a purposeful empty string property value.

Reply:
The `None` thing still gets me because if value is `None` the `json.loads` is guaranteed to raise an exception.  Which you then pass on and immediately check if `value is None`.  You could avoid the guaranteed exception if you reorder it.  Maybe some explicit returns instead of reassigning value a bunch would help to?

```python
    value = getattr(config, setting_name, None)

    if value is None:
        return getattr(settings, setting_name, default_value)

    try:
        return json.loads(value)
    except Exception:
        return value
```

Something like that.  Mostly the guaranteed exception path is what's bothering me.

Reply:
Updated it to this ordering.